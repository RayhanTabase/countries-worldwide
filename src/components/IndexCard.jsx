import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { countries } from 'country-flag-icons'

console.log(countries);

const IndexCard = ({ id, name, population }) => (
  <li>
    <Link to={`/details/${id}`}>
      <p>
        {name}
      </p>
      <p>
        {population}
      </p>
    </Link>
  </li>
);

IndexCard.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  population: PropTypes.string.isRequired,
  // flagName: PropTypes.string.isRequired,
};

export default IndexCard;
