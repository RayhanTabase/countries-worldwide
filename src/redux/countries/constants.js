export const START_REQUEST_COUNTRIES = 'START_REQUEST_COUNTRIES';
export const GET_COUNTRIES_SUCCESS = 'GET_CURRENT_BOOKS_SUCCESS';
export const GET_COUNTRIES_FAILURE = 'GET_COUNTRIES_FAILURE';
