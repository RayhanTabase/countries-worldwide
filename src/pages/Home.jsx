import { useState } from 'react';
import { useSelector } from 'react-redux';
import IndexCard from '../components/IndexCard';

const Home = () => {
  const [queryName, setQueryName] = useState('');
  const [queryRegion, setQueryRegion] = useState('all');
  const [pageNumber, setPageNumber] = useState(1);

  let { countries } = useSelector((state) => state.countriesReducer);
  const countriesLength = countries.length;
  const { loading } = useSelector((state) => state.countriesReducer);
  let regions = new Set(countries.map((country) => country.un_geoscheme.region));
  regions = [...regions];

  if (queryRegion !== 'all') {
    countries = countries.filter((country) => {
      const countryRegion = country.un_geoscheme.region;
      return countryRegion === queryRegion;
    });
  }

  if (queryName.trim()) {
    countries = countries.filter((country) => {
      const name = country.name.toUpperCase();
      return name.includes(queryName.trim().toUpperCase());
    });
  }

  countries = countries.slice((pageNumber - 1) * 6, pageNumber * 6);

  const changePageNumber = (value) => {
    if (value < 0 && pageNumber === 1) return;
    if (value > 0 && pageNumber * 6 > countriesLength) {
      return;
    }
    setPageNumber((prev) => prev + value);
  };

  return (
    <div>
      <input value={queryName} onChange={(e) => setQueryName(e.target.value)} placeholder="Search country by name" />
      <br />
      <select name="" value={queryRegion} onChange={(e) => setQueryRegion(e.target.value)}>
        <option value="all">All</option>
        { regions
          ? regions.map((region) => (
            <option key={region} value={region}>
              {region}
            </option>

          ))
          : ''}
      </select>

      { loading && <p>Loading...</p> }

      { countries.length < 1 && !loading
        ? (
          <p>Not found</p>
        )
        : (
          <ul>
            {countries.map((country) => (
              <IndexCard
                key={country.iso_3166.numeric}
                id={country.iso_3166.numeric}
                name={country.name}
                population={country.population.total}
                flagName={country.iso_3166.alpha2}
              />
            ))}
          </ul>
        )}
      <div>
        <button type="button" onClick={() => changePageNumber(-1)}>Prev</button>
        <p>{pageNumber}</p>
        <button type="button" onClick={() => changePageNumber(1)}>Next</button>
      </div>
    </div>
  );
};

export default Home;
