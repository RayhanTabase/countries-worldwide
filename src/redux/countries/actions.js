import { START_REQUEST_COUNTRIES, GET_COUNTRIES_SUCCESS, GET_COUNTRIES_FAILURE } from './constants';

export const startCountriesRequest = () => ({
  type: START_REQUEST_COUNTRIES,
});

export const getCountriesSuccess = (payload) => ({
  type: GET_COUNTRIES_SUCCESS,
  payload,
});

export const getCountriesFailure = (payload) => ({
  type: GET_COUNTRIES_FAILURE,
  payload,
});
