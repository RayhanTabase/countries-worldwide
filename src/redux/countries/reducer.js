import { START_REQUEST_COUNTRIES, GET_COUNTRIES_SUCCESS, GET_COUNTRIES_FAILURE } from './constants';

const initialState = {
  loading: true,
  countries: [],
  error: '',
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case START_REQUEST_COUNTRIES:
      return {
        ...state,
        loading: true,
      };

    case GET_COUNTRIES_SUCCESS:
      return {
        ...state,
        loading: false,
        countries: action.payload,
      };

    case GET_COUNTRIES_FAILURE:
      return {
        ...state,
        loading: false,
        error: `${action.payload}`,
      };

    default:
      return state;
  }
};

export default reducer;
