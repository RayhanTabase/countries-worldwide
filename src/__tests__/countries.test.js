import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import store from '../redux/configureStore';

import countryReducer from '../redux/countries/reducer';
import { START_REQUEST_COUNTRIES, GET_COUNTRIES_SUCCESS, GET_COUNTRIES_FAILURE } from '../redux/countries/constants';
import Home from '../pages/Home';
import IndexCard from '../components/IndexCard';
import data from '../utils/data';

describe('Home page test', () => {
  test('snapshot test for Home', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Home />
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Index card', () => {
  test('snapshot test for IndexCard component', () => {
    const { asFragment } = render(
      <IndexCard key="1" name="Ghana" population="3000" />,
      { wrapper: MemoryRouter },
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test('should render the correct text content', () => {
    const { container } = render(
      <IndexCard key="1" name="Ghana" population="3000" />,
      { wrapper: MemoryRouter },
    );
    const displayedText = container.querySelectorAll('p');
    expect(displayedText.length).toBe(2);
    expect(screen.getByText('Ghana')).toBeInTheDocument();
    expect(screen.getByText('3000')).toBeInTheDocument();
  });
});

describe('Unit test for countries reducer', () => {
  it('test that when API is still being fetched loading is set to true', () => {
    expect(countryReducer({ countries: [], loading: false, error: '' }, { type: START_REQUEST_COUNTRIES })).toEqual(
      {
        countries: [],
        loading: true,
        error: '',
      },
    );
  });

  it('test that data is fetched from the API and loading is set to false', () => {
    expect(countryReducer({ countries: [], loading: true, error: '' }, { type: GET_COUNTRIES_SUCCESS, payload: data })).toEqual(
      {
        countries: data,
        loading: false,
        error: '',
      },
    );
  });

  it('test that failed fetch', () => {
    const errorMessage = 'failed to respond';
    expect(countryReducer({ countries: [], loading: true, error: '' }, { type: GET_COUNTRIES_FAILURE, payload: errorMessage })).toEqual(
      {
        countries: [],
        loading: false,
        error: errorMessage,
      },
    );
  });
});
