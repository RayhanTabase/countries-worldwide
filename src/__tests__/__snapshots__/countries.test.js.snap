// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page test snapshot test for Home 1`] = `
<DocumentFragment>
  <div>
    <input
      placeholder="Search country by name"
      value=""
    />
    <br />
    <select
      name=""
    >
      <option
        value="all"
      >
        All
      </option>
    </select>
    <p>
      Loading...
    </p>
    <ul />
    <div>
      <button
        type="button"
      >
        Prev
      </button>
      <p>
        1
      </p>
      <button
        type="button"
      >
        Next
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Index card snapshot test for IndexCard component 1`] = `
<DocumentFragment>
  <li>
    <a
      href="/details/undefined"
    >
      <p>
        Ghana
      </p>
      <p>
        3000
      </p>
    </a>
  </li>
</DocumentFragment>
`;
